common/src/sql/dbinit.sql

```diff
@@ -281,37 +281,39 @@ CREATE UNIQUE INDEX ON omicron.public.silo_user (
 
 CREATE TABLE omicron.public.silo_group (
     id UUID PRIMARY KEY,
-    name STRING(128) NOT NULL,
-    description STRING(512) NOT NULL,
     time_created TIMESTAMPTZ NOT NULL,
     time_modified TIMESTAMPTZ NOT NULL,
     time_deleted TIMESTAMPTZ,
 
-    silo_id UUID NOT NULL
+    silo_id UUID NOT NULL,
+    external_id TEXT NOT NULL
 );
 
-CREATE UNIQUE INDEX ON omicron.public.silo_group (
-    name,
+CREATE INDEX ON omicron.public.silo_group (
```

Comment:
If I remove it, I see a full table scan error for

```
DELETE  FROM "silo_group_membership" WHERE ("silo_group_membership"."silo_group_id" = ANY(SELECT "silo_group"."id" FROM "silo_group" WHERE (("silo_group"."silo_id" = $1) AND ("silo_group"."time_deleted" IS NOT NULL))))
```

meaning it's coming from the silo delete code:

```rust
        let updated_rows =
            diesel::delete(silo_group_membership::dsl::silo_group_membership)
                .filter(
                    silo_group_membership::dsl::silo_group_id.eq_any(
                        silo_group::dsl::silo_group
                            .filter(silo_group::dsl::silo_id.eq(id))
                            .filter(silo_group::dsl::time_deleted.is_not_null())
                            .select(silo_group::dsl::id),
                    ),
                )
                .execute_async(self.pool_authorized(opctx).await?)
                .await
                .map_err(|e| {
                    public_error_from_diesel_pool(
                        e,
                        ErrorHandler::NotFoundByResource(authz_silo),
                    )
                })?;
```

but there's also the silo group delete code:

```rust
                diesel::update(dsl::silo_group)
                    .filter(dsl::id.eq(group_id))
                    .filter(dsl::time_deleted.is_null())
                    .set(dsl::time_deleted.eq(Utc::now()))
                    .execute(conn)?;
```

which also would use this index.

Reply:
Makes sense.  When those queries become paginated in the future, they're going to need the `id` column.  I'd suggest making this a `UNIQUE` index on `(silo_id, id)`.  I don't think it makes any difference right now.  But we've done that elsewhere for pagination and it'll avoid having to do a schema migration later.

Reply:
Right now, the indexes on silo_group are:

```sql
CREATE INDEX ON omicron.public.silo_group (
    silo_id
) WHERE
    time_deleted IS NOT NULL;

CREATE UNIQUE INDEX ON omicron.public.silo_group (
    silo_id,
    external_id
) WHERE
    time_deleted IS NULL;
```

With the column name change, I'm a little confused - are you suggesting changing the first one to `UNIQUE` on `(silo_id, id)`, or `(silo_id, external_id)`?

Reply:
Sorry I think I missed a few things here, especially that the `time_deleted` condition is inverted between these two indexes.  This index here is on `time_deleted IS NOT NULL`, so it includes only *deleted* silo groups.

I see the usage you mentioned in `DataStore::silo_delete()`.  That query indeed specifies `time_deleted IS NOT NULL`, so it matches this index.  But it looks to me like both might be wrong?  At that point in `silo_delete()`, we haven't deleted the silo groups yet, right?  That looks like the next thing we do.  So it seems like the "delete all silo group memberships" section will only delete memberships from silo groups that were already deleted before the Silo was deleted.  (Sorry if I'm misreading the code again!)

If the right thing is to flip the `time_deleted` condition in the query, then you'd want to flip the condition in this index too, and then it really should be redundant with the other index.  So you should be able to just remove this one.  (If I'm misreading the code and we _do_ want to filter only on the deleted ones, then I'd suggest instead dropping the `time_deleted IS NULL` condition on the other index and removing this one.)

The second usage you mentioned (silo group delete code) is specifying `time_deleted IS NULL`, so it cannot use this index (but should use the other index).

Reply:
Yep, that looks like a bug. We delete silo group memberships *before* silo groups, meaning `time_deleted` *is* null. I have no idea how this passes the restriction on full table scans...?

The only theory that makes sense to me is there wasn't a test for deleting a silo that contained a user, group, and group memberships, meaning the number of rows was zero. When I added one, the test failed because group memberships remained, not because of the full table scan restriction.

Commit 6e79986f changes the index, the filter in `silo_delete`, and adds the test I'm talking about.

Reply:
We discussed the indexes again offline and I'll summarize here.  I'd suggested just one unique, partial index here.  This triggered an unexpected error on the `INSERT ... ON CONFLICT`.  I filed #1545 to describe this and our subsequent research.  In conclusion I think there are two okay options here:

1. Keep the single unique partial index and use `on_conflict_do_nothing()`.  (@jmpesp I only discovered this option after we talked about this.)  We may want to add a test for trying to ensure the same group twice?
2. Keep the single unique partial index and drop the `on_conflict` part of the insert.  The problem with this is that it means that if two users log in at about the same time, both in some IdP group that we've never seen before, one of them will fail spuriously.  But if that user retries, it will work.  That's a lot of conditions, the impact is small, and there's a trivial workaround (retry).  It'd be nice to fix this but it's hard for me to justify asking to spend much time on this case right now.

Either is okay with me -- but we should do one of these.  As it is right now, since we have a unique non-partial index, it would incorrectly prevent us from reusing the external_id of a deleted silo group.

This is my last worry about this change -- otherwise I think it's good to go!

Reply:
@smklein found [another possible option](https://github.com/oxidecomputer/omicron/issues/1545#issuecomment-1205609289)

Reply:
Updated in df376126, thanks for debugging with me :)